### PR TITLE
Add LoongArch support

### PR DESCRIPTION
**Pull request by Rabenda**
### Motivation and Context

LoongArch is a new RISC ISA, which is a bit like MIPS or RISC-V.
Add LoongArch support

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by fantix** [comment]
Failed to build kernel modules, environment:
 * kernel 6.0.0-rc4-2 ([ArchLinux](https://mirrors.wsyu.edu.cn/loongarch/2022.09/) on QEMU 7.1)
 * glibc 2.36-4
 * gcc 12.2.1 20221009


```
  MODPOST /build/zfs-linux-git/src/zfs/module/Module.symvers
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'vm_map_base'
ERROR: modpost: GPL-incompatible module zfs.ko uses GPL-only symbol 'empty_zero_page'
```

```console
$ grep vm_map_base * -wrn
grep: module/os/linux/zfs/abd_os.o: binary file matches
grep: module/os/linux/zfs/zfs_znode.o: binary file matches
grep: module/os/linux/zfs/zfs_vnops_os.o: binary file matches
grep: module/os/linux/zfs/zfs_uio.o: binary file matches
```

```console
$ grep empty_zero_page * -wrn
grep: module/os/linux/zfs/abd_os.o: binary file matches
```

Possible relevant: #13199, but:

```console
...
checking whether ZERO_PAGE() is GPL-only... no
...
$ cat build/zero_page_license/build.log
make: Entering directory '/usr/lib/modules/6.0.0-rc4-2/build'
warning: the compiler differs from the one used to build the kernel
  The kernel was built by: gcc (GCC) 12.2.1 20220905
  You are using:           gcc (GCC) 12.2.1 20221009
  MODPOST /build/zfs-linux-git/src/zfs/build/zero_page_license/Module.symvers
  CC [M]  /build/zfs-linux-git/src/zfs/build/zero_page_license/zero_page_license.mod.o
  LD [M]  /build/zfs-linux-git/src/zfs/build/zero_page_license/zero_page_license.ko
make: Leaving directory '/usr/lib/modules/6.0.0-rc4-2/build'
$ cat build/zero_page_license/Module.symvers
$
```

**Review by ryao — COMMENTED** [review]
I would prefer if this LoongArch architecture support were implemented in a single commit. Other than that, this looks good to me.

Please squash the commits, rebase on the current master branch and repush.

**Review by ryao — APPROVED** [review]

File: include/os/linux/spl/sys/isa_defs.h [hunk 1]
```diff
@@ -223,10 +223,34 @@
...
+#if !defined(__loongarch64)
+#define	__loongarch64
+#endif
+
+#define	_ZFS_LITTLE_ENDIAN
+#define	_SUNOS_VTOC_16
+
+#ifndef _LP64
+#define	_LP64
+#endif
```
Comment (xen0n) [inline include/os/linux/spl/sys/isa_defs.h:242]: better check for `__loongarch_lp64`, as the `__loongarch64` macro [is deprecated](https://loongson.github.io/LoongArch-Documentation/LoongArch-toolchain-conventions-EN.html). You could fold the LP64 check into the `__loongarch__` check just like the riscv case.

File: include/os/linux/spl/sys/isa_defs.h [hunk 2]
```diff
@@ -223,10 +223,34 @@
...
+
+/*
+ * Illumos doesn't define _ALIGNMENT_REQUIRED for loongarch, so default to 1
+ * out of paranoia.
+ */
```
Comment (xen0n) [inline include/os/linux/spl/sys/isa_defs.h:247]: There isn't a LoongArch port of Illumos (yet), so better reword the comment. The actual reason why we really want to require alignment is that LoongArch implementations are explicitly not required to support unaligned accesses by the ISA spec.

File: include/os/linux/spl/sys/isa_defs.h [hunk 3]
```diff
@@ -223,10 +223,34 @@
 
 #define	_ALIGNMENT_REQUIRED	1
 
+/*
+ * loongarch arch specific defines
```
Comment (xen0n) [inline include/os/linux/spl/sys/isa_defs.h:227]: Use consistent/official spelling: "LoongArch"

File: module/lua/ldo.c [hunk 4]
```diff
@@ -84,6 +84,8 @@ static intptr_t stack_remaining(void) {
 #define JMP_BUF_CNT	18
 #elif defined(__riscv)
 #define JMP_BUF_CNT     64
+#elif defined(__loongarch64)
```
Comment (xen0n) [inline module/lua/ldo.c:87]: `__loongarch__` or `__loongarch_lp64`, pick one? Similarly migrate all other `__loongarch64` usages.

File: module/lua/setjmp/setjmp_loongarch64.S [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  .type symbol, @function; \
+  symbol:
+
+#define END(function) \
+  .size function, .- function;
+
+ENTRY(setjmp)
+	st.d $ra, $a0, 0*8
+	st.d $sp, $a0, 1*8
+	st.d $x,  $a0, 2*8
```
Comment (xen0n) [inline module/lua/setjmp/setjmp_loongarch64.S:42]: Please use `$r21` instead. The name `x` is deprecated for so long [it isn't even present in the official psABI doc from day one](https://github.com/loongson/LoongArch-Documentation/pull/3).

File: module/lua/setjmp/setjmp_loongarch64.S [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+	st.d $s2, $a0, 6*8
+	st.d $s3, $a0, 7*8
+	st.d $s4, $a0, 8*8
+	st.d $s5, $a0, 9*8
+	st.d $s6, $a0, 10*8
+	st.d $s7, $a0, 11*8
+	st.d $s8, $a0, 12*8
+
+	li.w		$v0, 0
+	jirl		$zero,$ra,0
```
Comment (xen0n) [inline module/lua/setjmp/setjmp_loongarch64.S:55]: Please use modern aliases (at least `jr $ra` is supported for every upstream binutils with LoongArch support, if not jumping to `ret` directly). No need to cater for ancient vendor binutils that doesn't support aliases. Also please don't use `$vX` as they're deprecated too.

`move $a0, $zero` + `jr $ra` should do it.

**Comment by mcmilk** [comment]
@Rabenda - would be nice, if this pull request could be updated. Maybe with the changes that @xen0n has suggested?

@fantix - Thank you for testing. Maybe you can re-test it, when the request got some rebase.

**Comment by xen0n** [comment]
I've just helped rebasing this and incorporate my review comments. Also, Linux v6.3 has the previously needed but GPL-only symbols (`empty_zero_page` and `vm_map_base`) relaxed to `EXPORT_SYMBOL`, so in theory this should now Just Work™️. (I've already heard multiple people have successfully mounted their pools from various user groups.)

**Review by behlendorf — APPROVED** [review]

**Pull request merged**